pep-0586.rst

```diff
@@ -0,0 +1,737 @@
+
+   from typing import Literal
+   from lib_with_no_types import SomeEnum  # SomeEnum has type 'Any'!
+
+   # x has type `Literal[Any]` due to the bad import
+   x: Literal[SomeEnum.FOO]
+
+Because ``Literal`` may not be parameterized by ``Any``, this program
+is *illegal*: the type checker should report an error with the last line.
+
+In short, while ``Any`` may effectively be used as a placeholder for any
+arbitrary *type*, it is currently **not** allowed to serve as a placeholder
+for any arbitrary *value*.
+
+Type inference
+==============
+
+This section describes a few rules regarding type inference and
+literals, along with some examples.
+
+Backwards compatibility
+-----------------------
+
+When type checkers add support for Literal, it's important they do so
+in a way that preserves backwards-compatibility. Code that used to
+type check **must** continue to type check after support for Literal
+is added.
+
+This is particularly important when performing type inference. For
+example, given the statement ``x = "blue"``, should the inferred
+type of ``x`` be ``str`` or ``Literal["blue"]``?
+
+This PEP does not require any particular strategy for cases like this,
+apart from requiring that backwards compatibility is maintained.
+
+For example, one simple strategy for meeting this requirement would be
+to always assume expressions are *not* Literal types unless they are
+explicitly annotated otherwise. A type checker using this strategy would
+always infer that ``x`` is of type ``str`` in the above example.
+
+If type checkers choose to use more sophisticated inference strategies,
+they should avoid being too over-zealous while doing so.
+
+For example, one strategy that does *not* work is always assuming expressions
+are Literal types. This naive strategy would cause programs like the
+following to start failing when they previously did not::
+
+    # If a type checker infers 'var' has type Literal[3]
+    # and my_list has type List[Literal[3]]...
+    var = 3
+    my_list = [var]
+
+    # ...this call would be a type-error.
+    my_list.append(4)
+
+Another example of when this strategy would fail is when setting fields
+in objects::
+
+    class MyObject:
+        def __init__(self) -> None:
+            # If a type checker infers MyObject.field has type Literal[3]...
+            self.field = 3
+
+    m = MyObject()
+
+    # ...this assignment would no longer type check
+    m.field = 4      
+
+Using non-Literals in Literal contexts
+--------------------------------------
+
+Literal types follow the existing rules regarding subtyping with no additional
+special-casing. For example, programs like the following are type safe::
+
+   def expects_str(x: str) -> None: ...
+   var: Literal["foo"] = "foo"
+
+   # Legal: Literal["foo"] is a subtype of str
+   expects_str(var)
+
+This also means non-Literal expressions in general should not automatically
+inferred to be Literal. For example::
```

Comment:
Just a wording nit: the previous section uses "infer" in a technical sense, so I think it's confusing to use it in a colloquial sense here. Maybe "should not be automatically cast to Literal" or "should not be considered subtypes of Literal"?

Reply:
Thanks! I made the fix you suggested here: https://github.com/python/peps/pull/929